Simplify exception chaining and nested exception error messages.

jamadden opened the pull request:
(Based on #40 since it touches many of the same things.)

Fixes #43

Given a/b/c/d.zcml included in that order with an error in d.zcml, previously executing `python -c "from zope.configuration.xmlconfig import file; file('/tmp/a.zcml')"` would get this:

```
Traceback (most recent call last):
   ...
    raise InvalidDottedName(value).with_field_and_value(self, value)
zope.schema.interfaces.InvalidDottedName: invalid dotted name

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
   ...
    raise InvalidDottedName(value).with_field_and_value(self, value)
zope.configuration.exceptions.ConfigurationError: ('Invalid value for', 'package', 'invalid dotted name')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
   ...
    raise InvalidDottedName(value).with_field_and_value(self, value)
zope.configuration.xmlconfig.ZopeXMLConfigurationError: File "/tmp/d.zcml", line 2.4-2.58
    ConfigurationError: ('Invalid value for', 'package', 'invalid dotted name')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
   ...
    raise InvalidDottedName(value).with_field_and_value(self, value)
zope.configuration.xmlconfig.ZopeXMLConfigurationError: File "/tmp/c.zcml", line 2.4-2.29
    ZopeXMLConfigurationError: File "/tmp/d.zcml", line 2.4-2.58
    ConfigurationError: ('Invalid value for', 'package', 'invalid dotted name')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
   ...
    raise InvalidDottedName(value).with_field_and_value(self, value)
zope.configuration.xmlconfig.ZopeXMLConfigurationError: File "/tmp/b.zcml", line 2.4-2.29
    ZopeXMLConfigurationError: File "/tmp/c.zcml", line 2.4-2.29
    ZopeXMLConfigurationError: File "/tmp/d.zcml", line 2.4-2.58
    ConfigurationError: ('Invalid value for', 'package', 'invalid dotted name')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  ...
    raise InvalidDottedName(value).with_field_and_value(self, value)
zope.configuration.xmlconfig.ZopeXMLConfigurationError: File "/tmp/a.zcml", line 2.4-2.29
    ZopeXMLConfigurationError: File "/tmp/b.zcml", line 2.4-2.29
    ZopeXMLConfigurationError: File "/tmp/c.zcml", line 2.4-2.29
    ZopeXMLConfigurationError: File "/tmp/d.zcml", line 2.4-2.58
    ConfigurationError: ('Invalid value for', 'package', 'invalid dotted name')
```

Now we get the simpler:

```
Traceback (most recent call last):
  ...
  File "/Users/jmadden/Projects/GithubSources/zope.schema/src/zope/schema/_field.py", line 670, in _validate
    raise InvalidDottedName(value).with_field_and_value(self, value)
zope.schema.interfaces.InvalidDottedName: invalid dotted name

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  ...
    raise InvalidDottedName(value).with_field_and_value(self, value)
zope.configuration.exceptions.ConfigurationError: Invalid value for 'package': InvalidDottedName('invalid dotted name')
    zope.schema.interfaces.InvalidDottedName: invalid dotted name
    File "/tmp/d.zcml", line 2.4-2.58
    File "/tmp/c.zcml", line 2.4-2.29
    File "/tmp/b.zcml", line 2.4-2.29
    File "/tmp/a.zcml", line 2.4-2.29
```

Yes, there's still a bit of repetition in that final case, but that seems OK to me?

This changed the constructors of two of the internal exceptions, but nobody should have been constructing those things anyway, right? 

Looking for feedback.

WIP because I need to do some tweaks for coverage.

 /cc @mgedmin

mgedmin commented:
> Yes, there's still a bit of repetition in that final case, but that seems OK to me?

But why?  The traceback looks the same, and the displayed exception message is the same.  What does the user gain from this duplication?

jamadden commented:
> But why? The traceback looks the same, and the displayed exception message is the same. What does the user gain from this duplication?

It's not *quite* the same. The extra info includes the module of the underlying exception, but the string doesn't. If that was all it is, I would drop the underlying exception from the top-level error message. But I think there's value in including most of the details on the single top-level line, mostly for line-oriented searches (possibly including some log analyzers).

I admit my arguments aren't especially strong here.

jamadden commented on src/zope/configuration/config.py:
```diff
@@ -769,29 +764,31 @@ def execute_actions(self, clear=True, testing=False):
+                            raise
+                        if not isinstance(ex, Exception):
+                            # BaseException
+                            raise
```
I was doing this in one catch block to avoid duplicating the `if isinstance(ex, ConfigurationError): ex.append_details(info); raise` logic. ConfigurationError could be caught by both `pass_through_exceptions` and `Exception`.

But I see now that I could just list `except ConfigurationError` *first* and solve the problem.

jamadden commented on src/zope/configuration/xmlconfig.py:
```diff
@@ -80,41 +79,40 @@ class ZopeXMLConfigurationError(ConfigurationError):
         else:
-            return str(v)
+            info = str(exception)
+        super(ZopeSAXParseException, self).__init__(info, exception)
```
TBH, I don't actually see the point of this exception anymore. My emacs is able to follow links to 'foo.xml:12:3' just fine, and it's not like we consistently reformat the parser location information. 

Maybe we can just drop this reformatting?

jamadden replied:
In fact, when I load malformed ZCML in a compile buffer, only the *original* form of the error is highlighted by emacs as something it can jump to. The reformat we're doing here is not.

mgedmin replied:
I suspect whoever wrote this code had their emacs configured to parse Python traceback lines.

I'm +1 for dropping the reformatting.

jamadden commented:
>> But why? The traceback looks the same, and the displayed exception message is the same. What does the user gain from this duplication?

During the process of updating the *other* place that could raise a very similar exception, I changed my mind and dropped the duplication.

jamadden commented:
@mgedmin How does this look now?

mgedmin commented:
I'm sure I'm going to love it, but it's a 300-line diff and today's a Saturday.

jamadden commented:
Thank you!
